### PR TITLE
Make cancel transaction idempotent

### LINKED ISSUE #15541: Make createCancelTransaction action idempotent

### PR DESCRIPTION
**Pull request by jpuri**
Fixes: #15541

Related to PR: https://github.com/MetaMask/metamask-extension/pull/15667

Make action createCancelTransaction idempotent.

### CODE REVIEW

**Review by brad-decker — DISMISSED** [review]

File: app/scripts/controllers/transactions/index.js [hunk 1]
```diff
@@ -1151,13 +1151,28 @@ export default class TransactionController extends EventEmitter {
...
+    actionId,
   ) {
+    // In transaction is found for same action id, do not create a new cancel transaction.
+    if (actionId) {
+      const existingTxMeta =
+        this.txStateManager.getTransactionWithActionId(actionId);
+      if (existingTxMeta) {
+        // Method approveTransaction is confusing, but it actually approves and also submits / publishes the transaction
+        if (existingTxMeta.status === TRANSACTION_STATUSES.APPROVED) {
+          await this.approveTransaction(existingTxMeta.id);
```
Comment (kumavis) [inline app/scripts/controllers/transactions/index.js:1170]: seems like this is added out of an abundance of caution but wanted to verify the state flow

https://github.com/MetaMask/metamask-extension/blob/2a27421e8e38612df47981faa6d59065cc810c2c/app/scripts/controllers/transactions/index.js#L1795-L1807

File: app/scripts/controllers/transactions/index.js [hunk 2]
```diff
@@ -1151,13 +1151,28 @@ export default class TransactionController extends EventEmitter {
...
    * @param options
    * @param options.estimatedBaseFee
+   * @param actionId
    * @returns {txMeta}
    */
   async createCancelTransaction(
     originalTxId,
     customGasSettings,
     { estimatedBaseFee } = {},
+    actionId,
```
Comment (digiwand) [inline app/scripts/controllers/transactions/index.js:1161]: could we combine this param with the options param above?

File: app/scripts/controllers/transactions/index.js (hunk 1 above)
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js]: @kumavis : PR is updated please check

File: app/scripts/controllers/transactions/index.js [hunk 3]
```diff
@@ -1293,7 +1305,11 @@ export default class TransactionController extends EventEmitter {
     // we need to keep track of what is currently being signed,
     // So that we do not increment nonce + resubmit something
     // that is already being incremented & signed.
-    if (this.inProcessOfSigning.has(txId)) {
+    const txMeta = this.txStateManager.getTransaction(txId);
+    if (
+      this.inProcessOfSigning.has(txId) ||
+      txMeta.status === TRANSACTION_STATUSES.SUBMITTED
+    ) {
```
Comment (jpuri) [inline app/scripts/controllers/transactions/index.js:1312]: The check is added to ensure that the fn does not re-approve a transaction `txMeta.status === TRANSACTION_STATUSES.SUBMITTED`.

File: app/scripts/controllers/transactions/index.js [hunk 4]
```diff
@@ -1161,13 +1161,24 @@ export default class TransactionController extends EventEmitter {
...
+   * @param actionId
    * @returns {txMeta}
    */
   async createCancelTransaction(
     originalTxId,
     customGasSettings,
     { estimatedBaseFee } = {},
+    actionId,
   ) {
+    // In transaction is found for same action id, do not create a new cancel transaction.
```
Comment (segun) [inline app/scripts/controllers/transactions/index.js:1173]:
```suggestion
    // If transaction is found for same action id, do not create a new cancel transaction.
```

**Review by segun — DISMISSED** [review]
LGTM

File: app/scripts/controllers/transactions/index.js [hunk 5]
```diff
@@ -1161,13 +1161,24 @@ export default class TransactionController extends EventEmitter {
...
    * @param options
    * @param options.estimatedBaseFee
+   * @param actionId
    * @returns {txMeta}
    */
   async createCancelTransaction(
     originalTxId,
     customGasSettings,
     { estimatedBaseFee } = {},
+    actionId,
```
Comment (brad-decker) [inline app/scripts/controllers/transactions/index.js:1171]: I think the same feedback that @digiwand left on the other pr applies here, IE this should either be a part of the options bag object, or moved to ahead of the options bag.

**Review by brad-decker — CHANGES_REQUESTED** [review]
Same feedback as the other PR regarding argument position

File: app/scripts/controllers/transactions/index.js (hunk 5 above)
Comment (jpuri) [reply app/scripts/controllers/transactions/index.js]: I updated the PR to implement feedback.

**Review by brad-decker — APPROVED** [review]
LGTM

**Review by digiwand — APPROVED** [review]

**Pull request merged**